ifaddrs.go

```diff
@@ -236,10 +236,14 @@ func IfAttrs(selectorName string, ifAddrs IfAddrs) (string, error) {
 	return attrVal, err
 }
 
-// GetAllInterfaces iterates over all available network interfaces and finds all
+// GetAllInterfaces proxies the reference to getAllInterfaces in order to
+// allow stubbing in tests by temporarily swapping implementations.
+var GetAllInterfaces = getAllInterfaces
+
+// getAllInterfaces iterates over all available network interfaces and finds all
 // available IP addresses on each interface and converts them to
 // sockaddr.IPAddrs, and returning the result as an array of IfAddr.
-func GetAllInterfaces() (IfAddrs, error) {
+func getAllInterfaces() (IfAddrs, error) {
 	ifs, err := net.Interfaces()
```

Comment:
to aid the set up of the "fake environment", we were curious if we could try to do some dependency injection type thing here and mock out subsequent calls to `net.Interfaces()`.

Does that make sense? Happy to chat more as needed

Reply:
Hi @FFMMM, yes it makes sense, I hope I fully understood what you meant, I would have asked for clarification but given the holidays I gave it a shot by introducing `NetworkInterfacesProvider` (naming is hard), its default implementation called `OSNetProvider` (that I thought made sense to move into a separate file but I may be wrong) and I also attempted in improving the way the "fake environment" in tests is configured.

One thing I don't particularly like is the fact that `sockaddr.NetProvider` is public, I would fix that by using a separate file with a public accessor and a build flag that would make it available only when testing.

If I missed the point, please clarify how you would like to inject the dependency that takes care of enumerating the net interfaces and I will adjust accordingly.